source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -3566,96 +3600,66 @@ void Graph::addNodePopup(bool cursor)
-                        std::string name = it->second[j][0];
                         if (ImGui::MenuItem(getUserNodeDefName(name).c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
                         {
-                            addNode(it->second[j][2], getUserNodeDefName(name), it->second[j][1]);
+                            addNode(node[2], getUserNodeDefName(name), node[1]);
                             _addNewNode = true;
                         }
                     }
-                    ImGui::EndMenu();
-                }
-            }
-        }
-
-        // Filter nodedefs and add to menu if matches filter
-        for (std::unordered_map<std::string, std::vector<mx::NodeDefPtr>>::iterator it = _nodesToAdd.begin(); it != _nodesToAdd.end(); ++it)
-        {
-            // Filter out list of nodes
-            if (subs.size() > 0)
-            {
-                ImGui::SetNextWindowSizeConstraints(ImVec2(250.0f, 300.0f), ImVec2(-1.0f, 500.0f));
-                for (size_t i = 0; i < it->second.size(); i++)
-                {
-                    std::string str(it->second[i]->getName());
-                    std::string nodeName = it->second[i]->getName();
-                    if (str.find(subs) != std::string::npos)
-                    {
-                        std::string val = getUserNodeDefName(nodeName);
-                        if (ImGui::MenuItem(val.c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
-                        {
-                            addNode(it->second[i]->getNodeString(), val, it->second[i]->getType());
-                            _addNewNode = true;
-                            memset(input, '\0', sizeof(input));
-                        }
-                    }
-                }
-            }
-            else
-            {
-                ImGui::SetNextWindowSizeConstraints(ImVec2(100, 10), ImVec2(-1, 300));
-                if (ImGui::BeginMenu(it->first.c_str()))
-                {
-                    ImGui::SetWindowFontScale(_fontScale);
-                    for (size_t i = 0; i < it->second.size(); i++)
+                    else
                     {
-                        std::string name = it->second[i]->getName();
-                        std::string val = getUserNodeDefName(name);
-                        if (ImGui::MenuItem(val.c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
+                        if (ImGui::BeginMenu(node[2].c_str()))
                         {
-                            addNode(it->second[i]->getNodeString(), val, it->second[i]->getType());
-                            _addNewNode = true;
+                            if (ImGui::MenuItem(getUserNodeDefName(name).c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
+                            {
+                                addNode(node[2], getUserNodeDefName(name), node[1]);
```

Comment:
The magic numbers to access array elements in the code is a bit hard to read. Maybe some named constants would help readability e.g. what does it mean to access position 2 in the array

Reply:
Would it be reasonable to create a new class here? coming from C# it pained me to write this code with all this weird array indexing and sort of faking an object. It seems like creating a `MenuItem` class with properties of name, type, group, and category would clean up both the vec<vec<string>> into a vec<MenuItem> and this weird indexing. 

not sure if there's an obvious place for that to live or if there are any gotchas / reasons to not do this

Reply:
That sounds like the right approach to me, @beersandrew, adding a `MenuItem` class that's local to the Graph Editor (e.g. in Graph.cpp or Graph.h), and leveraging this instead of a standard vector.

Reply:
Added the `MenuItem` class to clean this up